phasespace/fromdecay/genmultidecay.py

```diff
@@ -0,0 +1,290 @@
+        rand_i = tf.random.categorical(
+            tnp.log([[dm[0] for dm in self.gen_particles]]), n_events
+        )
+        # Input to tf.unique_with_counts must be 1D
+        dec_indices, _, counts = tf.unique_with_counts(rand_i[0])
+        counts = tf.cast(counts, tf.int64)
+        weights, max_weights, events = [], [], []
+        for i, n in zip(dec_indices, counts):
+            weight, max_weight, four_vectors = self.gen_particles[i][1].generate(
+                n, normalize_weights=False, **kwargs
+            )
+            weights.append(weight)
+            max_weights.append(max_weight)
+            events.append(four_vectors)
+
+        if normalize_weights:
+            total_max = tnp.max([tnp.max(mw) for mw in max_weights])
+            normed_weights = [w / total_max for w in weights]
+            return normed_weights, events
+
+        return weights, max_weights, events
+
+
+def _unique_name(name: str, preexisting_particles: set[str]) -> str:
+    """Create a string that does not exist in preexisting_particles based on name.
+
+    Args:
+        name: Original name
+        preexisting_particles: Names that the particle cannot have as name.
+
+    Returns:
+        name: Will be `name` if `name` is not in preexisting_particles or of the format "name [i]" where i
+            begins at 0 and increases until the name is not preexisting_particles.
+    """
+    if name not in preexisting_particles:
+        preexisting_particles.add(name)
+        return name
+
+    name += " [0]"
+    i = 1
+    while name in preexisting_particles:
+        name = name[: name.rfind("[")] + f"[{str(i)}]"
+        i += 1
+    preexisting_particles.add(name)
+    return name
+
+
+def _get_particle_mass(
+    name: str,
+    mass_converter: dict[str, Callable],
+    mass_func: str,
+    tolerance: float = _MASS_WIDTH_TOLERANCE,
+) -> Union[Callable, float]:
+    """
+    Get mass or mass function of particle using the particle package.
+    Args:
+        name: Name of the particle. Name must be recognizable by the particle package.
+        tolerance : See _recursively_traverse
+
+    Returns:
+        A function if the mass has a width smaller than tolerance. Otherwise, return a constant mass.
+    TODO try to cache results for this function in the future for speedup.
+    """
+    particle = Particle.from_evtgen_name(name)
+
+    if particle.width <= tolerance:
+        return tf.cast(particle.mass, tf.float64)
+    # If name does not exist in the predefined mass distributions, use Breit-Wigner
+    return mass_converter[mass_func](mass=particle.mass, width=particle.width)
+
+
+def _recursively_traverse(
+    decaychain: dict,
+    mass_converter: dict[str, Callable],
+    preexisting_particles: set[str] = None,
+    tolerance: float = _MASS_WIDTH_TOLERANCE,
+) -> list[tuple[float, GenParticle]]:
+    """Create all possible GenParticles by recursively traversing a dict from decaylanguage.

```

Comment:
Can you always write "DecayLanguage"?

Reply:
Resolved in bb80371975aa2764273b20c79f1a30b56e226a87